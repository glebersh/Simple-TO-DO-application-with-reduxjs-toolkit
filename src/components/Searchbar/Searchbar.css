.searchbar__input {
  width: 100%;
  border: 1px solid black;
  outline: none;
  margin-top: 3em;
  padding: 3px 10px;
  text-indent: .7em;
}