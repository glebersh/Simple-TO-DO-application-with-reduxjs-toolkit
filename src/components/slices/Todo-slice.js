import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

export const fetchTodos = createAsyncThunk(
  'todo/fetchTodos',
  async function (_, { rejectWithValue }) {
    try {
      const response = await
        fetch('https://jsonplaceholder.typicode.com/todos?_limit=20');

      if (!response.ok) {
        throw new Error('Server error!');
      }

      const data = await response.json();

      return data;
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const deleteTodo = createAsyncThunk(
  'todo/deleteTodo',
  async function (id, { rejectWithValue, dispatch }) {
    try {
      const response = await fetch(`https://jsonplaceholder.typicode.com/todos/${id}`,
        {
          method: 'DELETE',
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      dispatch(deleteItem({ id }));
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const toggleCompleteState = createAsyncThunk(
  'todo/toggleCompleteState',
  async function (id, { rejectWithValue, dispatch, getState }) {
    const todoItem = getState().todoReducer.todo.find(todo => todo.id === id);
    try {
      const response = await fetch(`https://jsonplaceholder.typicode.com/todos/${id}`,
        {
          method: 'PATCH',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            completed: !todoItem.completed
          })
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      dispatch(toggleDone({ id }));
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const addItemAsync = createAsyncThunk(
  'todo/addItemAsync',
  async function (text, { rejectWithValue, dispatch }) {
    try {
      const newTask = {
        title: text,
        completed: false,
        userId: 1,
      };
      const response = await fetch(`https://jsonplaceholder.typicode.com/todos/`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(newTask),
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      const data = await response.json();
      console.log(data);
      dispatch(addItem(data));
    }
    catch (error) {
      return rejectWithValue(error.message);
    }
  }
);


const todoSlice = createSlice({
  name: 'todos',
  initialState: {
    todo: [
      { id: 1, title: 'eat', completed: false },
      { id: 2, title: 'sleep', completed: false },
      { id: 3, title: 'drink', completed: false },
      { id: 4, title: 'maybe drink again', completed: false },
      { id: 5, title: 'learn redux', completed: false },
      { id: 6, title: 'have fun', completed: false },
    ],
    loadingStatus: null,
    errorStatus: null,
  },
  reducers: {
    deleteItem(state, action) {
      state.todo = state.todo.filter(item => item.id !== action.payload.id);
    },

    addItem(state, action) {
      state.todo.push(action.payload);
    },

    toggleDone(state, action) {
      const neededItem = state.todo.find(item => item.id === action.payload.id);
      neededItem.completed = !neededItem.completed;
    },

  },
  extraReducers: {
    [fetchTodos.pending]: (state) => {
      state.loadingStatus = 'loading'
    },
    [fetchTodos.fulfilled]: (state, action) => {
      state.loadingStatus = 'resolved'
      state.todo = action.payload
    },
    [fetchTodos.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
    [deleteTodo.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
    [toggleCompleteState.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
    [addItemAsync.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
  },
});


export const { addItem,
  deleteItem,
  toggleDone } = todoSlice.actions;

export default todoSlice.reducer;