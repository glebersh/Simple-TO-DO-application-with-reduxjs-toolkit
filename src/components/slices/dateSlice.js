import { createSlice } from "@reduxjs/toolkit";
import dayjs from "dayjs";

const initialState = '';

const dateSlice = createSlice({
  name: 'filters',
  initialState,
  reducers: {
    changeDateFilter: (_, action) => action.payload
  },
});

export const { changeDateFilter } = dateSlice.actions;

export default dateSlice.reducer;