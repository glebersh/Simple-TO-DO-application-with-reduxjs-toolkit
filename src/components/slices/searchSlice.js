import { createSlice } from "@reduxjs/toolkit";

const initialState = '';

const searchSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    changeSearchFilter: (_, action) => action.payload
  },
});

export const { changeSearchFilter } = searchSlice.actions;

export default searchSlice.reducer;