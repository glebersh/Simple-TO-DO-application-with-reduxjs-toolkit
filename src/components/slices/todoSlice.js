import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

export const fetchTodos = createAsyncThunk(
  'todo/fetchTodos',
  async function (_, { rejectWithValue }) {
    try {
      const response = await
        fetch('http://localhost:3004/todos/');

      if (!response.ok) {
        throw new Error('Server error!');
      }

      const data = await response.json();

      return data;
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const deleteTodo = createAsyncThunk(
  'todo/deleteTodo',
  async function (id, { rejectWithValue, dispatch }) {
    try {
      const response = await fetch(`http://localhost:3004/todos/${id}`,
        {
          method: 'DELETE',
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      dispatch(deleteItem({ id }));
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const toggleCompleteState = createAsyncThunk(
  'todo/toggleCompleteState',
  async function (id, { rejectWithValue, dispatch, getState }) {
    const todoItem = getState().todoReducer.todo.find(todo => todo.id === id);
    try {
      const response = await fetch(`http://localhost:3004/todos/${id}`,
        {
          method: 'PATCH',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            completed: !todoItem.completed
          })
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      dispatch(toggleDone({ id }));
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const addItemAsync = createAsyncThunk(
  'todo/addItemAsync',
  async function ({ text, reminderDate }, { rejectWithValue, dispatch }) {
    try {
      const newTask = {
        title: text,
        completed: false,
        userId: 1,
        date: reminderDate,
        priority: 'Low',
      };
      const response = await fetch(`http://localhost:3004/todos/`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(newTask),
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      const data = await response.json();
      dispatch(addItem(data));
    }
    catch (error) {
      return rejectWithValue(error.message);
    }
  }
);

export const editTextAsync = createAsyncThunk(
  'todo/editTextAsync',
  async function ({ id, editedText }, { rejectWithValue, dispatch }) {
    try {
      const response = await fetch(`http://localhost:3004/todos/${id}`,
        {
          method: 'PATCH',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            title: editedText
          })
        }
      );
      if (!response.ok) {
        throw new Error('Item was not deleted due to an occurred error');
      }
      dispatch(editText({ id, editedText }));
    } catch (error) {
      return rejectWithValue(error.message);
    }
  }
);


const todoSlice = createSlice({
  name: 'todos',
  initialState: {
    todo: [],
    loadingStatus: null,
    errorStatus: null,
  },
  reducers: {
    deleteItem(state, action) {
      state.todo = state.todo.filter(item => item.id !== action.payload.id);
    },

    addItem(state, action) {
      state.todo.push(action.payload);
    },

    toggleDone(state, action) {
      const neededItem = state.todo.find(item => item.id === action.payload.id);
      neededItem.completed = !neededItem.completed;
    },

    editText(state, action) {
      const neededItem = state.todo.find(item => item.id === action.payload.id);
      neededItem.title = action.payload.editedText;
    },

    togglePriority(state, action) {
      const neededItem = state.todo.find(item => item.id === action.payload.id);
      neededItem.priority = action.payload.priority;
    },
  },
  extraReducers: {
    [fetchTodos.pending]: (state) => {
      state.loadingStatus = 'loading'
    },
    [fetchTodos.fulfilled]: (state, action) => {
      state.loadingStatus = 'resolved'
      state.todo = action.payload
    },
    [fetchTodos.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
    [deleteTodo.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
    [toggleCompleteState.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
    [addItemAsync.rejected]: (state, action) => {
      state.loadingStatus = 'rejected'
      state.errorStatus = action.payload;
    },
  },
});


export const { addItem,
  deleteItem,
  toggleDone,
  editText, togglePriority } = todoSlice.actions;

export default todoSlice.reducer;