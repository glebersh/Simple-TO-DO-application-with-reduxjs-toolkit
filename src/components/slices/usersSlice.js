import { createSlice } from "@reduxjs/toolkit";

const usersSlice = createSlice({
  name: 'users',
  initialState: {
    users: []
  },
  reducers: {
    addUser(state, action) {
      state.users.push({
        id: 1,
        login: action.payload.login,
        password: action.payload.password,
      });
    },
  },
});

export const { addUser } = usersSlice.actions;

export default usersSlice.reducer;