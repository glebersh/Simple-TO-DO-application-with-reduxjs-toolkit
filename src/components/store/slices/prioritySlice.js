import { createSlice } from "@reduxjs/toolkit";

const initialState = 'All';

const prioritySlice = createSlice({
  name: 'priority',
  initialState,
  reducers: {
    changePriorityFilter: (_, action) => action.payload,
  },
});

export const { changePriorityFilter } = prioritySlice.actions;

export default prioritySlice.reducer;