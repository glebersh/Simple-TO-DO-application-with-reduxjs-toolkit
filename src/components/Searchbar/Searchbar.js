import React, { useEffect, useState } from 'react'

import { Input, InputGroup, InputRightAddon } from '@chakra-ui/react';
import { SearchIcon } from '@chakra-ui/icons';

import { changeSearchFilter } from '../store/slices/searchSlice';

import './Searchbar.css';
import { useDispatch } from 'react-redux';

const Searchbar = () => {
  const dispatch = useDispatch();
  const [searchtext, setSearchText] = useState('');

  useEffect(() => {
    dispatch(changeSearchFilter(searchtext));
  }, [searchtext])


  return (
    <InputGroup mt='3em'>
      <Input id='search' icon={<SearchIcon />}
        placeholder="Search by task..."
        value={searchtext} onChange={(e) => setSearchText(e.target.value)}
        borderColor='rgb(122,122,122)'
      />
      <InputRightAddon children={<SearchIcon />} bgColor='transparent' borderColor='rgb(122,122,122)' />
    </InputGroup>
  )
};

export default Searchbar;