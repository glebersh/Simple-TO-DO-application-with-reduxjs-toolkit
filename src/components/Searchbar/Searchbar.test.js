import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as reduxHooks from 'react-redux';
import Searchbar from "./Searchbar";
import * as actions from '../store/slices/searchSlice';

jest.mock('../store/slices/searchSlice');
const changeSearchFilter = jest.spyOn(actions, 'changeSearchFilter');

jest.mock('react-redux');
const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');

describe('Searchbar tests', () => {
  it('Should render and match snapshot', () => {
    const dispatch = jest.fn();
    mockUseDispatch.mockReturnValue(dispatch);

    const searchbar = render(
      <Searchbar />);
    expect(searchbar).toMatchSnapshot();
  });

  it('Should dispatch actions', () => {
    const dispatch = jest.fn();
    mockUseDispatch.mockReturnValue(dispatch);

    const searchbar = render(
      <Searchbar />);

    userEvent.type(searchbar.getByRole('searchbar'), 'Text to search');

    expect(dispatch).toHaveBeenCalledTimes(15);
    expect(changeSearchFilter).toHaveBeenCalledWith("Text to search");
  });
});
