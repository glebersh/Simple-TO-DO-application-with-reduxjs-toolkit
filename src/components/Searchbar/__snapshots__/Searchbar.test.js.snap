// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar tests Should render and match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="chakra-input__group css-d4ficv"
      >
        <input
          class="chakra-input css-117fu1l"
          icon="[object Object]"
          id="search"
          placeholder="Search by task..."
          role="searchbar"
          value=""
        />
        <div
          class="chakra-input__right-addon css-u3ali5"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-input__group css-d4ficv"
    >
      <input
        class="chakra-input css-117fu1l"
        icon="[object Object]"
        id="search"
        placeholder="Search by task..."
        role="searchbar"
        value=""
      />
      <div
        class="chakra-input__right-addon css-u3ali5"
      >
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
